tox_docker/tox4/config.py

```diff
@@ -0,0 +1,189 @@
+from collections import defaultdict
+from typing import Callable, Container, Dict, List, Sequence
+
+from tox.config.loader.section import Section
+from tox.config.main import Config
+from tox.config.sets import ConfigSet
+from tox.tox_env.api import ToxEnv
+
+from tox_docker.config import (
+    ContainerConfig,
+    RunningContainers,
+    validate_link,
+    validate_port,
+    validate_volume,
+)
+
+# nanoseconds in a second; named "SECOND" so that "1.5 * SECOND" makes sense
+SECOND = 1000000000
+
+EnvRunningContainers = Dict[ToxEnv, RunningContainers]
+
+
+class MissingRequiredSetting(Exception):
+    pass
+
+
+def required(setting_name: str) -> Callable[[str], str]:
+    def require_value(val: str) -> str:
+        if not val:
+            raise MissingRequiredSetting(setting_name)
+        return val
+
+    return require_value
+
+
+class EnvDockerConfigSet(ConfigSet):
+    def register_config(self) -> None:
+        self.add_config(
+            keys=["docker"],
+            of_type=List[str],
```

Comment:
This should be IMHO of type List[DockerConfigSet]

Reply:
How would this work? I'd need a post-process function, but typing is such that it can't change types between what the generic code parses and what the post-process returns...

Reply:
Why do you need a post-process function? Though likely the factory method is your goto place, see
 https://github.com/tox-dev/tox/blob/rewrite/tests/plugin/test_plugin_custom_config_set.py#L24-L36

For what it's worth now that we have a factory method I am thinking to yank/remove the post-processing (as can be achieved with a factory).

Reply:
There's an issue with typing here -- the `factory` gets called once for each item in a multi-line setting; but the typing requires its output to be `List[whatever]`; as a result, you get a list of lists of things.

Reply:
That shouldn't be the case, the code will first transform the multiline string to list, and then apply the factory for each element.

Reply:
See https://github.com/tox-dev/tox/blob/rewrite/src/tox/config/loader/convert.py#L58-L59

Reply:
With this code

```python
class EnvDockerConfigSet(ConfigSet):
    def register_config(self) -> None:
        def build_docker_config_set(container_name: object) -> DockerConfigSet:
            assert isinstance(container_name, str)
            return self._conf.get_section_config(
                section=Section("docker", container_name),
                base=[],
                of_type=DockerConfigSet,
                for_env=None,
            )

        self.add_config(
            keys=["docker"],
            of_type=List[DockerConfigSet],
            default=[],
            desc="docker image configs to load",
            factory=build_docker_config_set,
        )
```

I get this mypy error:

```$ mypy --exclude tox3 tox_docker
tox_docker/tox4/config.py:124: error: Argument "factory" to "add_config" of "ConfigSet" has incompatible type "Callable[[object], DockerConfigSet]"; expected "Optional[Callable[[object], List[DockerConfigSet]]]"
Found 1 error in 1 file (checked 19 source files)```

You're right that the code works as expected, but the typing is wrong.

Reply:
Indeed 🤔 not sure how to fix the typing here 🤦 probably need to define some overloads for containers 🤔 for the add_config

Reply:
Can you create an issue for that, please?

Reply:
https://github.com/tox-dev/tox/issues/2240